File: server/feed10/src/inbound.ts
import * as nosql from './nosqldb'
import * as o from './outbound'
import * as p from './participant'
import { Client } from 'cassandra-driver'

export class InboundModel extends o.OutboundModel{
   readonly to: p.ParticipantModel
   constructor(from: p.ParticipantModel, to: p.ParticipantModel, occurred: Date, subject: string, story: string) {
      super(from, occurred, subject, story)
      this.to = to
   }
}

export class InboundService extends nosql.Repository {
   constructor(nosqlClient: Client) {
      super(nosqlClient)
   }
   public async get(id: number): Promise<InboundModel[]> {
      const query = 'select toTimestamp(occurred) as occurred, fromparticipantid, subject, story from Inbound where participantid = ? order by occurred desc'

      const results = await this.nosqlClient.execute(query, [id], {prepare: true})
      return results.rows.map((row) => {
         const fp = new p.ParticipantModel(row.fromparticipantid, '')
         const tp = new p.ParticipantModel(id, '')
         return new InboundModel(fp, tp, row.occurred, row.subject, row.story)
      })
   }
   public async save(i: InboundModel): Promise<InboundModel> {
      const cql = 'insert into Inbound (ParticipantID, FromParticipantID, Occurred, Subject, Story) values (?, ?, now(), ?, ?)';
      this.nosqlClient.execute(cql, [i.to.id, i.from.id, i.subject, i.story], {prepare: true})
      return i
   }
}

File: server/feed13/src/feed/daos/inbound.clj
(ns feed.daos.inbound
  (:require [feed.daos.widecolumn :as wc])
  (:import (com.datastax.oss.driver.api.core CqlIdentifier)))

(def formatter (org.joda.time.format.DateTimeFormat/forPattern "yyyy-MM-dd"))

(defn convert-inbound
  "convert a row from the result to a map"
  [row id]
  (let [occurred-instant (.getInstant row (CqlIdentifier/fromCql "Occurred"))
        occurred-seconds (.getEpochSecond occurred-instant)
        occurred-millis (.toMillis (java.time.Duration/ofSeconds occurred-seconds))]
        {:from (.getInt row (CqlIdentifier/fromCql "FromParticipantID"))
         :to id
         :occurred (.print formatter occurred-millis)
         :subject (.getString row (CqlIdentifier/fromCql "Subject"))
         :story (.getString row (CqlIdentifier/fromCql "Story"))}))

(defn fetch
  "fetch the inbound news feed items for a participant"
  [id]
  (let [cql (str "select toTimestamp(occurred) as Occurred, FromParticipantID, Subject, Story from Inbound where participantid = " id " order by occurred desc")
        s (.execute @wc/cassandra cql)]
        (map #(convert-inbound % id) (.all s))))

(defn create
  "create an inbound news feed item for a participant"
  [from to occurred subject story]
  (let [cleansed-subject (.replaceAll subject "'" "")
        cleansed-story (.replaceAll story "'" "")
        cql (str "insert into Inbound (ParticipantID, FromParticipantID, Occurred, Subject, Story) values (" to ", " from ", now(), '" cleansed-subject "', '" cleansed-story "') using ttl 7776000")]
        (.execute @wc/cassandra cql))
  {:from from :to to :occurred occurred :subject subject :story story})
